### DISCUSSION TITLE
UI comparison in PR using github action

### DISCUSSION THREAD

**Discussion by synox** [adobe/helix-home#339] [Ideas]
### Story 

As a developer, I want to be sure that my PR does not break anything major on an unrelated page. 

### Idea

When a Pull Request created, a bot automatically compares the `main` with the `branch` rendering of each page based on a list of the most important pages and the [block library](https://github.com/adobe/franklin-sidekick-library). 

There is no need for storing historical screenshots, as we can load the main and branch pages at any time and directly compare them. 

> A visual difference has been detected for the following pages. To acknowledge/ignore, press the checkbox next to it. 
> [ ] /product/a {link to diff image}
> [ ] /block-library/teaser {link to diff image}

Block the PR until all the checkboxes are checked. 

### How
Ideally, an existing solution can be customized to fit this need. Or Playwright and Github actions can be used to compare screenshots.

**Comment by dylandepass** [comment]
I like the idea, I think if visual regression testing becomes a common ask from customers than using helix bot would be a nice experience.

Worth noting that if customers really need this and they use storybook then they can use [Chromatic](https://www.chromatic.com/) today with Franklin.. For example, it would only be a couple lines of code to get visual regression testing on [Rosalind](https://main--rosalind-boilerplate--dylandepass.hlx.page/) which is setup with storybook for blocks and templates.

**Comment by synox** [comment]
I created a simple PoC, see [PR](https://github.com/hlxsites/vg-volvotrucks-us/pull/412/files). It compares a list of paths, and also fetches a list of indexes to get more paths. 

Here is an example output after comparing 34 paths: https://github.com/hlxsites/vg-volvotrucks-us/pull/412#issuecomment-1536944669

![Screenshot 2023-05-05 at 17 55 46](https://user-images.githubusercontent.com/2250964/236589673-8aba3bfc-9c76-4918-8370-9bfaa4b807a4.png)

See also https://github.com/hlxsites/24life/pull/198

![Screenshot 2023-10-10 at 19 14 04](https://github.com/adobe/helix-home/assets/2250964/c75f35f6-79de-49fd-add0-20cb50cbdf25)

**Comment by sdmcraft** [comment]
We are using a variant of the playwright based visual tests suggested by @synox above for a customer project and it is working reasonable well for us. Though couple of feedback points:
1. It tends to make the PR conversations a bit too noisy if there are lot of syncs happening in it. @dnbute recently made a [few enhancements](https://github.com/hlxsites/sunstar/commit/64861b700e5031214c1a53dad1f11c0290acc09a) in the associated Github workflow to reuse the same comment.
2. As the number of test paths keep getting added, the duration of the visual tests keeps increasing. And these may not cover situation where I may want to run the tests against a broader/customized set of webpages on one-off basis. To this end, I extracted this in a separate standalone tool to run the visual tests. [More details can be found here](https://github.com/sdmcraft/web-tools/blob/main/docs/VISUAL-COMPARE.md).

**Reply by synox** [reply]
Thank you for this valuable feedback. More hands-on experience is definitely needed to make this tool more broadly usable.

**Reply by synox** [reply]
In additionally to the diff screenshots, I was thinking of forking https://github.com/pianomister/diffsite and using it to display pages side-by-side. This view is often easier to understand than the diff images.
